src/pam_hbac_ldap_compat.c

```diff
@@ -0,0 +1,297 @@
+/*
+    Copyright (C) 2016 Jakub Hrozek <jakub.hrozek@posteo.se>
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "pam_hbac_compat.h"
+#include "pam_hbac_ldap.h"
+
+#include <ctype.h>
+
+#if !defined(HAVE_LDAP_STR2DN) || defined(COMPAT_LDAP_UNIT_TESTS)
+static int rdn_keyval(const char *exploded_rdn_value,
+                      char **_key, char **_val)
+{
+    size_t i;
+    char *eq_sign;
+    const char *rdn_start;
+    const char *rdn_end;
+    char *val_start;
+    char *key = NULL;
+    char *value = NULL;
+    char *val_end = NULL;
+
+    if (exploded_rdn_value == NULL) {
+        return EINVAL;
+    }
+
+    /* Skip leading whitespace before the key */
+    rdn_start = exploded_rdn_value;
+    for (i = 0; isspace(rdn_start[i]); i++);
+
+    rdn_start += i;
+    if (*rdn_start == '\0') {
+        return EINVAL;
+    }
+
+    eq_sign = strchr(rdn_start, '=');
+    if (eq_sign == NULL) {
+        return EINVAL;
+    }
+
+    /* Skip whitespace before the '=' sign and after the key */
+    for (rdn_end = eq_sign;
+         isspace(*rdn_end);
+         rdn_end--);
+
+    val_start = eq_sign+1;
+    /* Skip leading whitespace before the value */
+    for (i =0; isspace(val_start[i]); i++);
+
+    /* Empty value? */
+    val_start += i;
+    if (*val_start == '\0') {
+        return EINVAL;
+    }
+
+    /* Strip trailing val whitespace */
+    for (val_end = val_start + strlen(val_start);
+         isspace(*val_end);
+         val_end--);
+
+    key = strndup(rdn_start, rdn_end - rdn_start);
+    value = strdup(val_start);
```

Comment:
On Mon, May 23, 2016 at 11:26:31AM -0700, Pavel Reichl wrote:
> > +    /* Skip leading whitespace before the value */
> > +    for (i =0; isspace(val_start[i]); i++);
> > +
> > +    /* Empty value? */
> > +    val_start += i;
> > +    if (*val_start == '\0') {
> > +        return EINVAL;
> > +    }
> > +
> > +    /* Strip trailing val whitespace */
> > +    for (val_end = val_start + strlen(val_start);
> > +         isspace(*val_end);
> > +         val_end--);
> > +
> > +    key = strndup(rdn_start, rdn_end - rdn_start);
> > +    value = strdup(val_start);
> 
> Did you forget to change this line - to use val_end and val_start as we do for key?

Yes, I did. The patches I will push next will also include a test.